File: internal/game/game.go
package game

import (
	"fmt"
	"math"
	"math/rand"
	"time"
)

type GameState struct {
	PlayerName    string    `json:"player_name"`
	Wood          int       `json:"wood"`
	Beer          int       `json:"beer"`
	PrestigePoints int      `json:"prestige_points"`
	AxeLevel      int       `json:"axe_level"`
	TreeHP        int       `json:"tree_hp"`
	MaxTreeHP     int       `json:"max_tree_hp"`
	Workers       Workers   `json:"workers"`
	Upgrades      Upgrades  `json:"upgrades"`
	Stats         Stats     `json:"stats"`
	StartTime     time.Time `json:"start_time"`
	LastUpdate    time.Time `json:"last_update"`
}

type Workers struct {
	PtitLu  int `json:"ptit_lu"`
	Mathieu int `json:"mathieu"`
	Vico    int `json:"vico"`
}

type Upgrades struct {
	AutoClicker       int `json:"auto_clicker"`
	LumberjackSchool  int `json:"lumberjack_school"`
	BreweryBonus      int `json:"brewery_bonus"`
	GoldenAxe         int `json:"golden_axe"`
}

type Stats struct {
	TotalTreesChopped   int   `json:"total_trees_chopped"`
	TotalWoodGained     int   `json:"total_wood_gained"`
	TotalBeersConsumed  int   `json:"total_beers_consumed"`
	TotalClicks         int   `json:"total_clicks"`
	WorkersHired        int   `json:"workers_hired"`
	CriticalHits        int   `json:"critical_hits"`
	AchievementsUnlocked []string `json:"achievements_unlocked"`
	PrestigeCount       int   `json:"prestige_count"`
}

type Config struct {
	TreeMinHP        int     `json:"tree_min_hp"`
	TreeMaxHP        int     `json:"tree_max_hp"`
	WoodMin          int     `json:"wood_min"`
	WoodMax          int     `json:"wood_max"`
	CriticalChance   float64 `json:"critical_chance"`
	CriticalMultiplier float64 `json:"critical_multiplier"`
	BeerBonusPerBeer float64 `json:"beer_bonus_per_beer"`
	TargetBeers      int     `json:"target_beers"`
	PrestigeBeers    int     `json:"prestige_beers"`
}

type Game struct {
	State  *GameState
	Config *Config
	rand   *rand.Rand
}

func NewGame(playerName string) *Game {
	config := &Config{
		TreeMinHP:        2,
		TreeMaxHP:        6,
		WoodMin:          1,
		WoodMax:          3,
		CriticalChance:   0.1,
		CriticalMultiplier: 2.5,
		BeerBonusPerBeer: 0.015,
		TargetBeers:      420,
		PrestigeBeers:    1000,
	}
	
	now := time.Now()
	state := &GameState{
		PlayerName:     playerName,
		Wood:           0,
		Beer:           0,
		PrestigePoints: 0,
		AxeLevel:       1,
		TreeHP:         5,
		MaxTreeHP:      5,
		Workers:        Workers{},
		Upgrades:       Upgrades{},
		Stats:          Stats{AchievementsUnlocked: []string{}},
		StartTime:      now,
		LastUpdate:     now,
	}
	
	game := &Game{
		State:  state,
		Config: config,
		rand:   rand.New(rand.NewSource(time.Now().UnixNano())),
	}
	
	game.respawnTree()
	return game
}

func (g *Game) ChopTree() (int, bool, error) {
	if g.State.TreeHP <= 0 {
		return 0, false, fmt.Errorf("tree already destroyed")
	}
	
	g.State.Stats.TotalClicks++
	
	baseDamage := g.State.AxeLevel
	prestigeBonus := 1.0 + float64(g.State.PrestigePoints)*0.02
	damage := int(float64(baseDamage) * prestigeBonus)
	
	isCritical := g.rand.Float64() < g.getCriticalChance()
	if isCritical {
		damage = int(float64(damage) * g.getCriticalMultiplier())
		g.State.Stats.CriticalHits++
	}
	
	g.State.TreeHP -= damage
	woodGained := 0
	
	if g.State.TreeHP <= 0 {
		woodGained = g.harvestTree()
		g.respawnTree()
	}
	
	g.State.LastUpdate = time.Now()
	return woodGained, isCritical, nil
}

func (g *Game) harvestTree() int {
	baseWood := g.Config.WoodMin + g.rand.Intn(g.Config.WoodMax-g.Config.WoodMin+1)
	
	beerBonus := 1.0 + float64(g.State.Beer)*g.Config.BeerBonusPerBeer
	prestigeBonus := 1.0 + float64(g.State.PrestigePoints)*0.05
	breweryBonus := 1.0 + float64(g.State.Upgrades.BreweryBonus)*0.5
	
	totalBonus := beerBonus * prestigeBonus * breweryBonus
	woodGained := int(float64(baseWood) * totalBonus)
	
	g.State.Wood += woodGained
	g.State.Stats.TotalTreesChopped++
	g.State.Stats.TotalWoodGained += woodGained
	
	return woodGained
}

func (g *Game) respawnTree() {
	g.State.MaxTreeHP = g.Config.TreeMinHP + g.rand.Intn(g.Config.TreeMaxHP-g.Config.TreeMinHP+1)
	g.State.TreeHP = g.State.MaxTreeHP
}

func (g *Game) getCriticalChance() float64 {
	base := g.Config.CriticalChance
	goldenAxeBonus := float64(g.State.Upgrades.GoldenAxe) * 0.15
	return base + goldenAxeBonus
}

func (g *Game) getCriticalMultiplier() float64 {
	base := g.Config.CriticalMultiplier
	goldenAxeBonus := float64(g.State.Upgrades.GoldenAxe) * 1.0
	return base + goldenAxeBonus
}

func (g *Game) UpgradeAxe() error {
	cost := g.getAxeUpgradeCost()
	if g.State.Wood < cost {
		return fmt.Errorf("not enough wood: need %d, have %d", cost, g.State.Wood)
	}
	
	g.State.Wood -= cost
	g.State.AxeLevel++
	g.State.LastUpdate = time.Now()
	
	return nil
}

func (g *Game) getAxeUpgradeCost() int {
	return int(15 * math.Pow(1.3, float64(g.State.AxeLevel-1)))
}

func (g *Game) BuyWorker(workerType string) error {
	var cost int
	var currentCount *int
	
	switch workerType {
	case "ptit_lu":
		cost = g.getWorkerCost("ptit_lu", g.State.Workers.PtitLu)
		currentCount = &g.State.Workers.PtitLu
	case "mathieu":
		cost = g.getWorkerCost("mathieu", g.State.Workers.Mathieu)
		currentCount = &g.State.Workers.Mathieu
	case "vico":
		cost = g.getWorkerCost("vico", g.State.Workers.Vico)
		currentCount = &g.State.Workers.Vico
	default:
		return fmt.Errorf("unknown worker type: %s", workerType)
	}
	
	if g.State.Wood < cost {
		return fmt.Errorf("not enough wood: need %d, have %d", cost, g.State.Wood)
	}
	
	g.State.Wood -= cost
	*currentCount++
	g.State.Stats.WorkersHired++
	g.State.LastUpdate = time.Now()
	
	return nil
}

func (g *Game) getWorkerCost(workerType string, currentCount int) int {
	baseCosts := map[string]int{
		"ptit_lu": 25,
		"mathieu": 100,
		"vico":    300,
	}
	
	multipliers := map[string]float64{
		"ptit_lu": 1.6,
		"mathieu": 1.5,
		"vico":    1.35,
	}
	
	base := baseCosts[workerType]
	multiplier := multipliers[workerType]
	
	return int(float64(base) * math.Pow(multiplier, float64(currentCount)))
}

func (g *Game) BuyUpgrade(upgradeType string) error {
	var cost int
	var currentLevel *int
	
	switch upgradeType {
	case "auto_clicker":
		cost = g.getUpgradeCost("auto_clicker", g.State.Upgrades.AutoClicker)
		currentLevel = &g.State.Upgrades.AutoClicker
	case "lumberjack_school":
		cost = g.getUpgradeCost("lumberjack_school", g.State.Upgrades.LumberjackSchool)
		currentLevel = &g.State.Upgrades.LumberjackSchool
	case "brewery_bonus":
		cost = g.getUpgradeCost("brewery_bonus", g.State.Upgrades.BreweryBonus)
		currentLevel = &g.State.Upgrades.BreweryBonus
	case "golden_axe":
		cost = g.getUpgradeCost("golden_axe", g.State.Upgrades.GoldenAxe)
		currentLevel = &g.State.Upgrades.GoldenAxe
	default:
		return fmt.Errorf("unknown upgrade type: %s", upgradeType)
	}
	
	if g.State.Wood < cost {
		return fmt.Errorf("not enough wood: need %d, have %d", cost, g.State.Wood)
	}
	
	g.State.Wood -= cost
	*currentLevel++
	g.State.LastUpdate = time.Now()
	
	return nil
}

func (g *Game) getUpgradeCost(upgradeType string, currentLevel int) int {
	baseCosts := map[string]int{
		"auto_clicker":      75,
		"lumberjack_school": 500,
		"brewery_bonus":     200,
		"golden_axe":        1000,
	}
	
	multipliers := map[string]float64{
		"auto_clicker":      1.4,
		"lumberjack_school": 1.8,
		"brewery_bonus":     2.0,
		"golden_axe":        1.2,
	}
	
	base := baseCosts[upgradeType]
	multiplier := multipliers[upgradeType]
	
	return int(float64(base) * math.Pow(multiplier, float64(currentLevel)))
}

func (g *Game) BuyBeer() error {
	cost := g.getBeerCost()
	if g.State.Wood < cost {
		return fmt.Errorf("not enough wood: need %d, have %d", cost, g.State.Wood)
	}
	
	g.State.Wood -= cost
	g.State.Beer++
	g.State.Stats.TotalBeersConsumed++
	g.State.LastUpdate = time.Now()
	
	return nil
}

func (g *Game) getBeerCost() int {
	return int(4 * math.Pow(1.08, float64(g.State.Beer)))
}

func (g *Game) CanPrestige() bool {
	return g.State.Beer >= g.Config.PrestigeBeers
}

func (g *Game) Prestige() error {
	if !g.CanPrestige() {
		return fmt.Errorf("not enough beers for prestige: need %d, have %d", g.Config.PrestigeBeers, g.State.Beer)
	}
	
	prestigePointsGained := g.State.Beer / 100
	g.State.PrestigePoints += prestigePointsGained
	g.State.Stats.PrestigeCount++
	
	g.State.Wood = 0
	g.State.Beer = 0
	g.State.AxeLevel = 1
	g.State.Workers = Workers{}
	g.State.Upgrades = Upgrades{}
	g.respawnTree()
	g.State.LastUpdate = time.Now()
	
	return nil
}

func (g *Game) IsGameWon() bool {
	return g.State.Beer >= g.Config.TargetBeers
}

func (g *Game) ProcessWorkers() int {
	if time.Since(g.State.LastUpdate) < time.Second {
		return 0
	}
	
	workerEfficiency := map[string]float64{
		"ptit_lu": 0.8,
		"mathieu": 2.5,
		"vico":    8.0,
	}
	
	schoolBonus := 1.0 + float64(g.State.Upgrades.LumberjackSchool)*0.25
	prestigeBonus := 1.0 + float64(g.State.PrestigePoints)*0.02
	
	totalDamage := 0.0
	totalDamage += float64(g.State.Workers.PtitLu) * workerEfficiency["ptit_lu"]
	totalDamage += float64(g.State.Workers.Mathieu) * workerEfficiency["mathieu"]
	totalDamage += float64(g.State.Workers.Vico) * workerEfficiency["vico"]
	
	totalDamage *= schoolBonus * prestigeBonus
	
	if g.State.Upgrades.AutoClicker > 0 {
		autoClickerDamage := float64(g.State.Upgrades.AutoClicker) * 0.5
		totalDamage += autoClickerDamage
	}
	
	damage := int(totalDamage)
	woodGained := 0
	
	if damage > 0 {
		g.State.TreeHP -= damage
		if g.State.TreeHP <= 0 {
			woodGained = g.harvestTree()
			g.respawnTree()
		}
		g.State.LastUpdate = time.Now()
	}
	
	return woodGained
}

func (g *Game) GetPrices() map[string]int {
	return map[string]int{
		"axe_upgrade":       g.getAxeUpgradeCost(),
		"ptit_lu":          g.getWorkerCost("ptit_lu", g.State.Workers.PtitLu),
		"mathieu":          g.getWorkerCost("mathieu", g.State.Workers.Mathieu),
		"vico":             g.getWorkerCost("vico", g.State.Workers.Vico),
		"beer":             g.getBeerCost(),
		"auto_clicker":     g.getUpgradeCost("auto_clicker", g.State.Upgrades.AutoClicker),
		"lumberjack_school": g.getUpgradeCost("lumberjack_school", g.State.Upgrades.LumberjackSchool),
		"brewery_bonus":    g.getUpgradeCost("brewery_bonus", g.State.Upgrades.BreweryBonus),
		"golden_axe":       g.getUpgradeCost("golden_axe", g.State.Upgrades.GoldenAxe),
	}
}

func (g *Game) GetGameDuration() time.Duration {
	return time.Since(g.State.StartTime)
}

File: static/game.js
class ValouniversaireGame {
    constructor() {
        this.playerName = '';
        this.gameState = null;
        this.updateInterval = null;
        this.initializeElements();
        this.bindEvents();
        this.showLoading();
    }

    initializeElements() {
        this.elements = {
            loading: document.getElementById('loading'),
            gameContainer: document.getElementById('game-container'),
            gameArea: document.getElementById('game-area'),
            playerNameInput: document.getElementById('player-name'),
            startGameBtn: document.getElementById('start-game'),
            
            woodCount: document.getElementById('wood-count'),
            beerCount: document.getElementById('beer-count'),
            prestigeCount: document.getElementById('prestige-count'),
            gameDuration: document.getElementById('game-duration'),
            
            treeButton: document.getElementById('tree-button'),
            treeHpFill: document.getElementById('tree-hp-fill'),
            treeHpText: document.getElementById('tree-hp-text'),
            axeLevel: document.getElementById('axe-level'),
            upgradeAxe: document.getElementById('upgrade-axe'),
            axeCost: document.getElementById('axe-cost'),
            
            woodGained: document.getElementById('wood-gained'),
            criticalHit: document.getElementById('critical-hit'),
            
            buyBeer: document.getElementById('buy-beer'),
            beerCost: document.getElementById('beer-cost'),
            
            prestigeSection: document.getElementById('prestige-section'),
            prestigeButton: document.getElementById('prestige-button'),
            
            victoryScreen: document.getElementById('victory-screen'),
            continuePlayingBtn: document.getElementById('continue-playing'),
            resetGameBtn: document.getElementById('reset-game'),
            
            notificationContainer: document.getElementById('notification-container')
        };

        this.workerElements = {
            'ptit_lu': {
                button: document.getElementById('buy-ptit-lu'),
                count: document.getElementById('ptit-lu-count'),
                cost: document.getElementById('ptit-lu-cost')
            },
            'mathieu': {
                button: document.getElementById('buy-mathieu'),
                count: document.getElementById('mathieu-count'),
                cost: document.getElementById('mathieu-cost')
            },
            'vico': {
                button: document.getElementById('buy-vico'),
                count: document.getElementById('vico-count'),
                cost: document.getElementById('vico-cost')
            }
        };

        this.upgradeElements = {
            'auto_clicker': {
                button: document.getElementById('buy-auto-clicker'),
                count: document.getElementById('auto-clicker-count'),
                cost: document.getElementById('auto-clicker-cost')
            },
            'lumberjack_school': {
                button: document.getElementById('buy-lumberjack-school'),
                count: document.getElementById('lumberjack-school-count'),
                cost: document.getElementById('lumberjack-school-cost')
            },
            'brewery_bonus': {
                button: document.getElementById('buy-brewery-bonus'),
                count: document.getElementById('brewery-bonus-count'),
                cost: document.getElementById('brewery-bonus-cost')
            },
            'golden_axe': {
                button: document.getElementById('buy-golden-axe'),
                count: document.getElementById('golden-axe-count'),
                cost: document.getElementById('golden-axe-cost')
            }
        };

        this.statElements = {
            treesChopped: document.getElementById('trees-chopped'),
            totalWood: document.getElementById('total-wood'),
            totalBeers: document.getElementById('total-beers'),
            totalClicks: document.getElementById('total-clicks'),
            criticalHits: document.getElementById('critical-hits'),
            workersHired: document.getElementById('workers-hired')
        };
    }

    bindEvents() {
        this.elements.startGameBtn.addEventListener('click', () => this.startGame());
        this.elements.playerNameInput.addEventListener('keypress', (e) => {
            if (e.key === 'Enter') this.startGame();
        });

        this.elements.treeButton.addEventListener('click', () => this.chopTree());
        this.elements.upgradeAxe.addEventListener('click', () => this.upgradeAxe());
        this.elements.buyBeer.addEventListener('click', () => this.buyBeer());
        this.elements.prestigeButton.addEventListener('click', () => this.prestige());

        Object.keys(this.workerElements).forEach(workerType => {
            this.workerElements[workerType].button.addEventListener('click', () => this.buyWorker(workerType));
        });

        Object.keys(this.upgradeElements).forEach(upgradeType => {
            this.upgradeElements[upgradeType].button.addEventListener('click', () => this.buyUpgrade(upgradeType));
        });

        this.elements.continuePlayingBtn.addEventListener('click', () => this.hideVictoryScreen());
        this.elements.resetGameBtn.addEventListener('click', () => this.resetGame());
    }

    showLoading() {
        this.elements.loading.style.display = 'flex';
        this.elements.gameContainer.style.display = 'none';
    }

    hideLoading() {
        this.elements.loading.style.display = 'none';
        this.elements.gameContainer.style.display = 'block';
    }

    async startGame() {
        const playerName = this.elements.playerNameInput.value.trim();
        if (!playerName) {
            this.showNotification('Veuillez entrer votre nom', 'error');
            return;
        }

        this.playerName = playerName;
        this.hideLoading();
        this.elements.gameArea.style.display = 'block';
        
        await this.loadGameState();
        this.startUpdateLoop();
    }

    async loadGameState() {
        try {
            const response = await this.apiCall('GET', `/api/state?player_name=${encodeURIComponent(this.playerName)}`);
            if (response.success) {
                this.gameState = response.data;
                this.updateUI();
            }
        } catch (error) {
            this.showNotification('Erreur de chargement du jeu', 'error');
        }
    }

    async performAction(action, target = null) {
        try {
            const response = await this.apiCall('POST', '/api/action', {
                player_name: this.playerName,
                action: action,
                target: target
            });

            if (response.success) {
                this.gameState = response.data.game_state;
                this.updateUI();
                return response.data.action_result;
            } else {
                this.showNotification(response.error, 'error');
                return null;
            }
        } catch (error) {
            this.showNotification('Erreur lors de l\'action', 'error');
            return null;
        }
    }

    async chopTree() {
        const result = await this.performAction('chop');
        if (result) {
            this.animateTreeChop(result.wood_gained, result.is_critical);
        }
    }

    async upgradeAxe() {
        await this.performAction('upgrade_axe');
    }

    async buyWorker(workerType) {
        await this.performAction('buy_worker', workerType);
    }

    async buyUpgrade(upgradeType) {
        await this.performAction('buy_upgrade', upgradeType);
    }

    async buyBeer() {
        await this.performAction('buy_beer');
    }

    async prestige() {
        if (confirm('Êtes-vous sûr de vouloir faire un prestige? Cela remettra à zéro votre progression actuelle.')) {
            const result = await this.performAction('prestige');
            if (result) {
                this.showNotification('Prestige activé! Vous avez gagné des points de prestige permanents!');
            }
        }
    }

    async resetGame() {
        if (confirm('Êtes-vous sûr de vouloir recommencer le jeu?')) {
            try {
                const response = await this.apiCall('POST', `/api/reset?player_name=${encodeURIComponent(this.playerName)}`);
                if (response.success) {
                    this.gameState = response.data;
                    this.updateUI();
                    this.hideVictoryScreen();
                    this.showNotification('Jeu remis à zéro!');
                }
            } catch (error) {
                this.showNotification('Erreur lors de la remise à zéro', 'error');
            }
        }
    }

    updateUI() {
        if (!this.gameState) return;

        this.elements.woodCount.textContent = this.gameState.wood.toLocaleString();
        this.elements.beerCount.textContent = this.gameState.beer.toLocaleString();
        this.elements.prestigeCount.textContent = this.gameState.prestige_points.toLocaleString();
        this.elements.gameDuration.textContent = this.gameState.game_duration;

        const treeHpPercent = (this.gameState.tree_hp / this.gameState.max_tree_hp) * 100;
        this.elements.treeHpFill.style.width = `${treeHpPercent}%`;
        this.elements.treeHpText.textContent = `${this.gameState.tree_hp}/${this.gameState.max_tree_hp}`;

        this.elements.axeLevel.textContent = this.gameState.axe_level;
        this.elements.axeCost.textContent = this.gameState.prices.axe_upgrade.toLocaleString();

        this.elements.beerCost.textContent = this.gameState.prices.beer.toLocaleString();

        Object.keys(this.workerElements).forEach(workerType => {
            const workerCount = this.gameState.workers[workerType] || 0;
            this.workerElements[workerType].count.textContent = workerCount;
            this.workerElements[workerType].cost.textContent = this.gameState.prices[workerType].toLocaleString();
        });

        Object.keys(this.upgradeElements).forEach(upgradeType => {
            const upgradeLevel = this.gameState.upgrades[upgradeType] || 0;
            this.upgradeElements[upgradeType].count.textContent = upgradeLevel;
            this.upgradeElements[upgradeType].cost.textContent = this.gameState.prices[upgradeType].toLocaleString();
        });

        this.statElements.treesChopped.textContent = this.gameState.stats.total_trees_chopped.toLocaleString();
        this.statElements.totalWood.textContent = this.gameState.stats.total_wood_gained.toLocaleString();
        this.statElements.totalBeers.textContent = this.gameState.stats.total_beers_consumed.toLocaleString();
        this.statElements.totalClicks.textContent = this.gameState.stats.total_clicks.toLocaleString();
        this.statElements.criticalHits.textContent = this.gameState.stats.critical_hits.toLocaleString();
        this.statElements.workersHired.textContent = this.gameState.stats.workers_hired.toLocaleString();

        this.updateButtonStates();
        this.updatePrestigeSection();
        this.checkVictory();
    }

    updateButtonStates() {
        if (!this.gameState) return;

        const wood = this.gameState.wood;

        this.elements.upgradeAxe.disabled = wood < this.gameState.prices.axe_upgrade;
        this.elements.buyBeer.disabled = wood < this.gameState.prices.beer;

        Object.keys(this.workerElements).forEach(workerType => {
            const button = this.workerElements[workerType].button;
            const canAfford = wood >= this.gameState.prices[workerType];
            button.disabled = !canAfford;
            button.classList.toggle('affordable', canAfford);
        });

        Object.keys(this.upgradeElements).forEach(upgradeType => {
            const button = this.upgradeElements[upgradeType].button;
            const canAfford = wood >= this.gameState.prices[upgradeType];
            button.disabled = !canAfford;
            button.classList.toggle('affordable', canAfford);
        });
    }

    updatePrestigeSection() {
        if (!this.gameState) return;

        if (this.gameState.can_prestige) {
            this.elements.prestigeSection.style.display = 'block';
        } else {
            this.elements.prestigeSection.style.display = 'none';
        }
    }

    checkVictory() {
        if (this.gameState && this.gameState.is_game_won && this.elements.victoryScreen.style.display === 'none') {
            this.showVictoryScreen();
        }
    }

    showVictoryScreen() {
        this.elements.victoryScreen.style.display = 'flex';
    }

    hideVictoryScreen() {
        this.elements.victoryScreen.style.display = 'none';
    }

    animateTreeChop(woodGained, isCritical) {
        this.elements.treeButton.classList.add('animate');
        
        setTimeout(() => {
            this.elements.treeButton.classList.remove('animate');
        }, 600);

        if (isCritical) {
            this.elements.treeButton.classList.add('critical');
            this.elements.criticalHit.textContent = 'CRITIQUE!';
            this.elements.criticalHit.classList.add('show');
            
            setTimeout(() => {
                this.elements.criticalHit.classList.remove('show');
                this.elements.treeButton.classList.remove('critical');
            }, 1500);
        }

        if (woodGained > 0) {
            this.elements.woodGained.textContent = `+${woodGained} 🪵`;
            this.elements.woodGained.classList.add('show');
            
            setTimeout(() => {
                this.elements.woodGained.classList.remove('show');
            }, 1500);
        }
    }

    showNotification(message, type = 'success') {
        const notification = document.createElement('div');
        notification.className = `notification ${type}`;
        notification.textContent = message;
        
        this.elements.notificationContainer.appendChild(notification);
        
        setTimeout(() => {
            notification.classList.add('show');
        }, 100);
        
        setTimeout(() => {
            notification.classList.remove('show');
            setTimeout(() => {
                if (notification.parentNode) {
                    notification.parentNode.removeChild(notification);
                }
            }, 300);
        }, 3000);
    }

    startUpdateLoop() {
        this.updateInterval = setInterval(async () => {
            await this.loadGameState();
        }, 2000);
    }

    stopUpdateLoop() {
        if (this.updateInterval) {
            clearInterval(this.updateInterval);
            this.updateInterval = null;
        }
    }

    async apiCall(method, url, data = null) {
        const options = {
            method: method,
            headers: {
                'Content-Type': 'application/json'
            }
        };

        if (data) {
            options.body = JSON.stringify(data);
        }

        const response = await fetch(url, options);
        return await response.json();
    }
}

document.addEventListener('DOMContentLoaded', () => {
    window.game = new ValouniversaireGame();
});

window.addEventListener('beforeunload', () => {
    if (window.game) {
        window.game.stopUpdateLoop();
    }
});